### ISSUE TITLE
Consumer Lag for transactional message

### ISSUE THREAD

**Issue by ossberyl** [edenhill/librdkafka#2539]
Description
===========
I can see the consumer lag keeps to be 1 in rdkafka_consumer_example_cpp from even for master code base when it consumes from a Kafka Topic where the producer uses transaction.

It looks like the "[transaction control message](http://kafka.apache.org/documentation/#controlbatch)" at the last offset won't come to or is ignored by the rdkafka_consumer_example_cpp and never be committed back.

How to reproduce
================

- Produce message to test topic "TransactionTest" via transaction:

```
public class TransactionalProducer {
    public static void main(String[] args) {
        Properties props = new Properties();
        props.put("bootstrap.servers", "localhost:9092");
        props.put("transactional.id", "my-transactional-id");
        props.put("acks", "all");
        props.put("key.serializer", "org.apache.kafka.common.serialization.StringSerializer");
        props.put("value.serializer", "org.apache.kafka.common.serialization.StringSerializer");

        Producer<String, String> producer = new KafkaProducer<>(props, new StringSerializer(), new StringSerializer());

        producer.initTransactions();

        try {
            producer.beginTransaction();
            for (int i = 0; i < 100; i++)
                producer.send(new ProducerRecord<>("TransactionTest", Integer.toString(i), Integer.toString(i)));
            producer.commitTransaction();
        } catch (ProducerFencedException | OutOfOrderSequenceException | AuthorizationException e) {
            producer.close();
        } catch (KafkaException e) {
            producer.abortTransaction();
        }
        producer.close();
    }
}
```

- Use rdkafka_consumer_example_cpp to consume from above topic: (It's same result with or without setting "isolation.level"):
`./rdkafka_consumer_example_cpp -g test_librdkafka_master -b localhost:9092 -X topic.auto.offset.reset=smallest TransactionTest [-X isolation.level=read_committed]`

- Check consumer lag after all messages have been consumed by above consumer:
`kafka-consumer-groups.sh --bootstrap-server localhost:9092 --describe --group test_librdkafka_master`

```
TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG     
TransactionTest 16         9               10              1               
TransactionTest 4          5               6               1               
TransactionTest 14         3               4               1               
TransactionTest 18         6               7               1               
TransactionTest 17         4               5               1               
TransactionTest 7          8               9               1               
TransactionTest 6          3               4               1               
TransactionTest 15         7               8               1               
TransactionTest 13         7               8               1               
TransactionTest 12         5               6               1               
TransactionTest 3          7               8               1               
TransactionTest 2          2               3               1               
TransactionTest 0          3               4               1               
TransactionTest 1          3               4               1               
TransactionTest 5          2               3               1               
TransactionTest 11         4               5               1               
TransactionTest 19         6               7               1               
TransactionTest 8          9               10              1               
TransactionTest 10         6               7               1               
TransactionTest 9          1               2               1 
```

- But for a java kafka-console-consumer.sh, the lag would become 0.
`kafka-console-consumer.sh --bootstrap-server localhost:9092 --from-beginning --group testjava --topic TransactionTest`

`kafka-consumer-groups.sh --bootstrap-server localhost:9092 --describe --group testjava`

```
TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG
TransactionTest 16         10              10              0               
TransactionTest 4          6               6               0               
TransactionTest 14         4               4               0               
TransactionTest 18         7               7               0               
TransactionTest 17         5               5               0               
TransactionTest 7          9               9               0               
TransactionTest 6          4               4               0               
TransactionTest 15         8               8               0               
TransactionTest 13         8               8               0               
TransactionTest 12         6               6               0               
TransactionTest 3          8               8               0               
TransactionTest 2          3               3               0               
TransactionTest 0          4               4               0               
TransactionTest 1          4               4               0               
TransactionTest 5          3               3               0               
TransactionTest 11         5               5               0               
TransactionTest 19         7               7               0               
TransactionTest 8          10              10              0               
TransactionTest 10         7               7               0               
TransactionTest 9          2               2               0 
```              
             

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version (release number or git tag): `<master>`
 - [x] Apache Kafka version: `<2.0.0>`
 - [x] librdkafka client configuration: `<topic.auto.offset.reset=smallest>`

**Comment by mhowlett** [comment]
https://twitter.com/codinghorror/status/506010907021828096

**Comment by ossberyl** [comment]
Hi @mhowlett ,

Do we have any thoughts about this issue?

Thanks!

**Comment by mhowlett** [comment]
sorry for the delay @ossberyl - the behavior does indeed differ from the java client. As you suspect it's because offsets corresponding to control messages are not stored. PR opened to resolve this - #2661 - the fix should be in v1.4.
